File: src/main/java/ca/mcgill/science/tepid/client/CurrentUser.java
package ca.mcgill.science.tepid.client;

public class CurrentUser {
	
	public String user, domain;
	
	private CurrentUser() {
	}
	
	public static CurrentUser getCurrentUser() {
		CurrentUser cu = new CurrentUser();
		try {
			byte[] out = new byte[1024];
			Process p = new ProcessBuilder("whoami", "/upn").start();
			int bytes = p.getInputStream().read(out);
			p.getInputStream().close();
			String[] result = new String(out, 0, bytes).trim().split("@");
			cu.user = result[0];
			if (result.length > 1) cu.domain = result[1];
		} catch (Exception e) {
			try {
				byte[] out = new byte[1024];
				Process p = new ProcessBuilder("whoami").start();
				int bytes = p.getInputStream().read(out);
				p.getInputStream().close();
				String[] result = new String(out, 0, bytes).trim().split("\\\\");
				cu.user = result[0];
				if (result.length > 1) {
					cu.domain = result[0];
					cu.user = result[1];
				}
			} catch (Exception e1) {
				e.printStackTrace();
				e1.printStackTrace();
			}
		}
		return cu;
	}
	
	public static void main(String[] args) {
		System.out.println(getCurrentUser());
	}

	@Override
	public String toString() {
		return "CurrentUser [user=" + user + ", domain=" + domain + "]";
	}
	
}

File: src/main/java/ca/mcgill/science/tepid/client/models/CurrentUser.kt
package ca.mcgill.science.tepid.client.models

import ca.mcgill.science.tepid.utils.WithLogging
import java.io.InputStream

data class CurrentUser(val user: String, val domain: String?) {

    companion object : WithLogging() {

        val currentUser: CurrentUser? by lazy {
            try {
                return@lazy getCurrentUser(arrayOf("whoami", "/upn"), "@")
            } catch (e: Exception) {
                try {
                    return@lazy getCurrentUser(arrayOf("whoami"), "\\\\")
                } catch (e1: Exception) {
                    log.error("Failed to get user with upn", e)
                    log.error("Failed to get user with just whoami", e1)
                }
                return@lazy null
            }
        }

        private fun getCurrentUser(command: Array<String>, splitter: String): CurrentUser =
                ProcessBuilder(*command).start().inputStream.getCurrentUser(splitter)

        private fun InputStream.getCurrentUser(splitter: String): CurrentUser = use {
            val out = ByteArray(1024)
            val bytes = read(out)
            val userString = String(out, 0, bytes).trim()
            if (!userString.contains(splitter))
                return@use CurrentUser(userString, null)
            val (user, domain) = userString.split(splitter)
            return@use CurrentUser(user, domain)
        }
    }

}
